csm_web/scheduler/models.py

```diff
@@ -30,150 +35,118 @@ class Attendance(models.Model):
+        abstract = True
 
-    leader = models.ForeignKey(
-        "self",
+
+class Student(Profile):
+    section = models.ForeignKey(
+        "Section",
         on_delete=models.CASCADE,
-        related_name="followers",
         blank=True,
         null=True,
-    )
-    course = models.ForeignKey(Course, on_delete=models.CASCADE)
-    role = models.CharField(max_length=2, choices=ROLE_CHOICES, blank=True)
-    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
-    section = models.ForeignKey(
-        "Section", on_delete=models.CASCADE, blank=True, null=True
+        related_name="students",
     )
 
-    def clean(self):
-        if self.section:
-            mentor_set = self.section._get_mentor_set()
-            if mentor_set.count() > 1:
-                raise ValidationError(_("Cannot save section with more than 1 mentor!"))
-            if self.course != self.section.course:
-                raise ValidationError(_("Profile course must match section course!"))
+    class Meta:
+        unique_together = ("user", "section")
 
-    @property
-    def name(self):
-        return f"{self.user.first_name} {self.user.last_name}"
 
-    def __str__(self):
-        return f"{self.course.name} {self.role} - {self.name}"
+class Mentor(Profile):
+    pass
 
 
-class Section(models.Model):
+class Section(ValidatingModel):
     course = models.ForeignKey(Course, on_delete=models.CASCADE)
-    default_spacetime = models.OneToOneField("Spacetime", on_delete=models.CASCADE)
+    spacetime = models.OneToOneField("Spacetime", on_delete=models.CASCADE)
     capacity = models.PositiveSmallIntegerField()
-
-    def _get_mentor_set(self):
-        # excluding STUDENT doesn't work because of empty role
-        return self.profile_set.filter(
-            role__in=[
-                Profile.JUNIOR_MENTOR,
-                Profile.ASSOCIATE_MENTOR,
-                Profile.SENIOR_MENTOR,
-                Profile.COORDINATOR,
-            ]
-        )
-
-    @property
-    def students(self):
-        return self.profile_set.filter(role=Profile.STUDENT)
-
-    @property
-    def mentor(self):
-        mentor_profiles = mentor_profiles = self._get_mentor_set()
-        assert mentor_profiles.count() <= 1
-        return mentor_profiles.first()
+    mentor = models.ForeignKey(Mentor, on_delete=models.SET_NULL, blank=True, null=True)
 
     @property
     def current_student_count(self):
-        return self.active_students.count()
+        return self.students.count()
 
     current_student_count.fget.short_description = "Number of students enrolled"
 
-    @property
-    def leader(self):
-        return self.mentor
-
-    @property
-    def active_students(self):
-        return self.students.filter(active=True)
+    def clean(self):
+        super().clean()
+        if self.students.count() > self.capacity:
+            raise ValidationError(
+                "Number of enrolled students exceeds section capacity"
```

Comment:
there are use cases (albeit somewhat hacky ones) where we DO want to enroll over capacity
for example, to handle EOP sections, I set their max capacity to 0 and enrolled students manually: raising the capacity doesn't work, as if one of those students drops it would allow another non-EOP student to enroll through the site

instead, we should have this validation in serializer or views

Reply:
Interesting. Maybe we should consider adding a `public = BooleanField()` to `Section` instead of this hack? Thoughts?

Reply:
i think it's better to have some extra flexibility in this regard, since there may be more potential use cases (i think there was one other situation where it would have been helpful)

Reply:
Okay, will revert